File: go/sort/quick_sort.go
package sort

//快速排序

func QuickSort(data []int) {
	quickSortHelper(data, 0, len(data)-1)
}

func quickSortHelper(data []int, lo, hi int) {
	if lo < hi {
		mid := partition(data, lo, hi)
		quickSortHelper(data, lo, mid-1)
		quickSortHelper(data, mid+1, hi)
	}
}

func partition(data []int, lo, hi int) int {
	pivot, i, j := data[hi], lo, lo
	for j < hi {
		if data[j] < pivot {
			data[j], data[i] = data[i], data[j]
			i++
		}
		j++
	}
	data[i], data[hi] = data[hi], data[i]
	return i
}

// 快速排序-迭代实现
func QuickSortIteration(data []int) {
	stack := []int{0, len(data) - 1}

	for len(stack) > 1 {
		lo, hi := stack[len(stack)-2], stack[len(stack)-1]
		stack = stack[:len(stack)-2]
		if lo < hi {
			mid := partition(data, lo, hi)
			stack = append(stack, lo, mid-1)
			stack = append(stack, mid+1, hi)
		}
	}
}

File: python/sort/quick_sort.py
from typing import List


# 快速排序


def quick_sort(data: List[int]):
    quick_sort_helper(data, 0, len(data) - 1)


def quick_sort_helper(data, lo, hi):
    if lo >= hi:
        return
    mid = partition(data, lo, hi)
    quick_sort_helper(data, lo, mid - 1)
    quick_sort_helper(data, mid + 1, hi)


def partition(data, lo, hi) -> int:
    pivot = data[hi]
    i, j = lo, lo
    while j < hi:
        if data[j] < pivot:
            data[i], data[j] = data[j], data[i]
            i += 1
        j += 1
    data[i], data[hi] = data[hi], data[i]
    return i
